ballista/scheduler/src/circuit_breaker/controller.rs

```diff
@@ -0,0 +1,161 @@
+            },
+        );
+    }
+
+    pub fn delete(&self, job_id: &str) {
+        info!("Deleting circuit breaker for job {}", job_id);
+        self.job_states.remove(job_id);
+    }
+
+    pub async fn update(
+        &self,
+        task_identity: String,
+        partition: u32,
+        percent: f64,
+    ) -> Result<bool, String> {
+        let task_identity = Self::parse_task_identity(task_identity)?;
+
+        let stage_states = match self.job_states.get(&task_identity.job_id) {
+            Some(state) => state.stage_states.clone(),
+            // If the registration hasn't happened yet
+            None => {
+                warn!(
+                    "Circuit breaker update received for unregistered job {}",
+                    task_identity.job_id
+                );
+                return Ok(false);
+            }
+        };
+
+        let attempt_states = match stage_states.get(&task_identity.stage_id) {
+            Some(state) => state.attempt_states.clone(),
+            None => {
+                let attempt_states = Arc::new(DashMap::new());
+                stage_states.insert(
+                    task_identity.stage_id.clone(),
+                    StageState {
+                        attempt_states: attempt_states.clone(),
+                    },
+                );
+                attempt_states
+            }
+        };
+
+        let partition_states = match attempt_states.get(&task_identity.stage_attempt_num)
+        {
+            Some(state) => state.partition_states.clone(),
+            None => {
+                let partition_states = Arc::new(DashMap::new());
+                attempt_states.insert(
+                    task_identity.stage_attempt_num,
+                    AttemptState {
+                        partition_states: partition_states.clone(),
+                    },
+                );
+                partition_states
+            }
+        };
+
+        match partition_states.get_mut(&partition) {
+            Some(mut state) => {
+                state.percent = percent;
+            }
+            None => {
+                partition_states.insert(partition, PartitionState { percent });
+            }
+        };
+
+        let should_circuit_breaker = partition_states
+            .iter()
+            .map(|s| s.percent)
+            .fold(0.0, |a, b| a + b)
+            >= 1.0;
+
+        if should_circuit_breaker {
+            info!(
+                "Sending circuit breaker signal to task {}, global limit reached",
+                task_identity.to_key()
+            );
+        }
+
+        Ok(should_circuit_breaker)
+    }
+
+    fn parse_task_identity(task_identity: String) -> Result<TaskIdentity, String> {
+        // Ballista uses the same task identity format consistently, but it has to use the String defined by Datafusion.
+        // So we need to parse it here.
+        let regex = Regex::new(r"TID (\d+) ([^/]*)/(\d+)\.(\d+)/\[([\d,\s]+)\]").unwrap();
```

Comment:
nit: should put this in a `lazy_static` or initialize it when we create the controller so we don't have to compile the regex every time.